### PR TITLE
Tactic_option supports export

### LINKED ISSUE #15072: `Obligation Tactic` should support `#[export]` attribute, and should default to `#[export]` rather than `#[global]`
#### Description of the problem

```coq
Require Import Coq.Program.Program.
#[export] Obligation Tactic := idtac. (* Error: export attribute not supported here *)
```

Also
```coq
Require Import Coq.Program.Program.
Axiom admit : forall {T}, T.
Module foo.
  Obligation Tactic := exact admit.
End foo.

Program Definition bar : False := _.
Print bar. (* bar = bar_obligation_1
     : False *)
```

This is yet another example of side-effects of `Require` that are hard to banish.

#### Coq Version

8.13.1, also master (2880bc7ccd0edc47ccbfeee74acbb51ae768e293)

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #15072

Overlays:
- https://github.com/coq-community/paramcoq/pull/95

OUTDATED default is now a special backwards compatible locality
Note that there is no way to get a backwards compatible declaration
because of the following behaviour:

~~~coq
Module M.
  Module N.
    Obligation Tactic := idtac 1.
  End N.
End M.

Obligation Tactic := idtac 2.

Import M.

Show Obligation Tactic. (* "idtac 1" !!! *)
~~~

In other words, duplication to have both `#[global]` and `#[export]`
is as close as we can get but not backwards compatible.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
>Note that there is no way to get a backwards compatible declaration

A replacement (but not backwards compatible since export wasn't available) would be like
~~~coq
Module M.
  Module N.
    Module Inner.
      #[global] Obligation Tactic := idtac 1.
      #[export] Obligation Tactic := idtac 1.
    End Inner.
    Export Inner.
  End N.
  Export N.Inner.
End M.
Export M.N.Inner.
~~~

**Comment by Zimmi48** [comment]
Perhaps a strategy then would be to introduce `export` now but without changing the default behavior for now so that people can have a backward-compatible fix when the default behavior (the one of `global` IIUC) is fixed. OTOH, if it significantly complicates the implementation, that might not be worth it.

**Comment by SkySkimmer** [comment]
>Perhaps a strategy then would be to introduce export now but without changing the default behavior for now so that people can have a backward-compatible fix when the default behavior (the one of global IIUC) is fixed. OTOH, if it significantly complicates the implementation, that might not be worth it.

Let's see how it goes.

**Comment by SkySkimmer** [comment]
added changelog

File: plugins/ltac/tactic_option.ml [hunk 1]
```diff
@@ -8,40 +8,65 @@
...
+
+let tac_option_locality = Attributes.option_locality
+
+let warn_default_locality =
+  CWarnings.create ~name:"deprecated-tacopt-without-locality" ~category:"deprecated" Pp.(fun () ->
+      strbrk
+        "The default locality for this command outside sections is \
+         currently like the combination of \"global\", \"export\" and \
+         re-exporting for every surrounding module. It will change to \
+         \"export\" in a future release.")
```
Comment (Zimmi48) [inline plugins/ltac/tactic_option.ml:22]: Given the change to the meaning of `global`, adding the deprecation warning now is problematic because there is no backward-compatible way of fixing it. It seems that a better strategy would be to release this PR without the warning and to add the warning in the next release. BTW, I believe that most other warnings about the default locality do not mention as explicitly that the plan is for the new locality to be `export`, giving us some freedom to discuss making the default locality `local` instead (see part of the discussion at #13394).

Comment (SkySkimmer) [reply plugins/ltac/tactic_option.ml:22]: >Given the change to the meaning of global, adding the deprecation warning now is problematic because there is no backward-compatible way of fixing it. It seems that a better strategy would be to release this PR without the warning and to add the warning in the next release.

I don't understand what the point of delaying is.

Comment (Zimmi48) [reply plugins/ltac/tactic_option.ml]: Users usually do not like getting warnings that they cannot fix in backward-compatible ways. But maybe this is not such a big deal as users can choose between dropping compatibility with previous Coq releases, or disabling the warning until they do.

**Comment by SkySkimmer** [comment]
@coqbot ci minimize

**Comment by SkySkimmer** [comment]
Right, the meaning of Global changes.
~~~coq
Module X.
Obligation Tactic := idtac 1.
End X.
Module Import M.
Export X.
Global Obligation Tactic := idtac 2.
End M.
(* before: idtac 2, after: idtac 1 *)
~~~
because Export X is done after loading objects for M, and Global stops having an effect at import.

**Comment by Zimmi48** [comment]
Another option would be to keep the default locality and `global` in sync but introduce a compatibility flag that restores the previous behavior.

**Comment by SkySkimmer** [comment]
As discussed in coq call, global and default localities retain the old (backward compatible) behaviour but warn, and we support combined `#[global,export]` which doesn't warn.

File: plugins/ltac/tactic_option.ml [hunk 2]
```diff
@@ -9,39 +9,90 @@
...
+  | None, Some (), None -> return Global
+  | None, None, Some () -> return Export
+  | None, Some (), Some () -> return GlobalAndExport
+
+let warn_default_locality =
+  CWarnings.create ~name:"deprecated-tacopt-without-locality" ~category:"deprecated" Pp.(fun () ->
+      strbrk
+        "The default and global localities for this command outside \
+         sections are currently equivalent to the combination of \
+         \"global\" (with the meaning used by the \"Set\" command), \
```
Comment (Zimmi48) [inline plugins/ltac/tactic_option.ml:41]: Is this really necessary to explain the standard meaning of `global` by comparing it to another command? What about just say "the standard meaning of `global` as defined in the reference manual"? (https://coq.inria.fr/refman/language/core/modules.html#coq:attr.global)

File: plugins/ltac/tactic_option.ml [hunk 3]
```diff
@@ -9,39 +9,90 @@
...
+         \"export\" and re-exporting for every surrounding module. It \
+         will change to just \"global\" (with the meaning used by the \
+         \"Set\" command) in a future release." ++ fnl() ++
+      strbrk
+        "To preserve the current meaning, use the attribute \
```
Comment (Zimmi48) [inline plugins/ltac/tactic_option.ml:46]:
```suggestion
        "To preserve the current meaning in a future-proof way, use the attribute \
```

or:
```suggestion
        "To preserve the current meaning in a forward-compatible way, use the attribute \
```

**Comment by SkySkimmer** [comment]
changed warning wording

**Comment by SkySkimmer** [comment]
added overlay, doc

File: doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst [hunk 4]
```diff
@@ -0,0 +1,8 @@
+- **Changed:** commands setting tactic options (currently
```
Comment (jfehrle) [inline doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst:1]:
```suggestion
- **Changed:** commands that set tactic options (currently
```

File: doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+  :opt:`Firstorder Solver` and :cmd:`Obligation Tactic`, as well as any
+  defined by third party plugins) now support :attr:`export` locality.
```
Comment (jfehrle) [inline doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst:3]: This is somewhere between vague and circular: how do you know which commands defined by third party plugins are commands that set tactic options?  Perhaps you mean options/commands that have a value in the form `@ltac_expr` as the argument? Looks like `Obligation Tactic` should have been an option.  Is that worth cleaning up later?

"commands setting" is a bit loose since it also includes options.  Maybe "commands and options setting ..." is better?

File: doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+  Note that the meaning of :attr:`global` and the default outside
+  sections currently also set the option when importing any
+  surrounding module. This will change in a future version. (`#15274
```
Comment (jfehrle) [inline doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst:6]:
```suggestion
  Note that :attr:`global` and the default outside
  sections are currently applied when importing any
  surrounding module. This will change in a future version. (`#15274
```

The "surrounding module" part doesn't make immediate sense to me.  If this is already in the doc in a reasonable place, is it necessary to tersly explain it again here?

File: doc/sphinx/addendum/program.rst [hunk 7]
```diff
@@ -273,10 +273,17 @@ optional tactic is replaced by the default one if not specified.
...
-   This command supports the :attr:`local` and :attr:`global` attributes.
+   This command supports the :attr:`local`, :attr:`export` and :attr:`global` attributes.
    :attr:`local` makes the setting last only for the current
    module. :attr:`local` is the default inside sections while :attr:`global`
-   otherwise.
+   otherwise. :attr:`export` and :attr:`global` may be used together.
+
+   With no attribute and when :attr:`global` is not combined with
+   :attr:`export`, the meaning is different from the usual meaning of
+   :attr:`global`: the command's effect persists after the current
```
Comment (jfehrle) [inline doc/sphinx/addendum/program.rst:283]: "effect persists" singular or "effects persist" plural?

File: doc/sphinx/addendum/program.rst [hunk 8]
```diff
@@ -273,10 +273,17 @@ optional tactic is replaced by the default one if not specified.
...
    e.g. using :cmd:`Next Obligation`.
 
-   This command supports the :attr:`local` and :attr:`global` attributes.
+   This command supports the :attr:`local`, :attr:`export` and :attr:`global` attributes.
    :attr:`local` makes the setting last only for the current
    module. :attr:`local` is the default inside sections while :attr:`global`
-   otherwise.
+   otherwise. :attr:`export` and :attr:`global` may be used together.
+
+   With no attribute and when :attr:`global` is not combined with
```
Comment (jfehrle) [inline doc/sphinx/addendum/program.rst:281]:
```suggestion
   When no attribute is specified and :attr:`global` is not combined with
```

File: doc/sphinx/proofs/automatic-tactics/logic.rst [hunk 9]
```diff
@@ -124,8 +124,9 @@ Solvers for logic and equality
 
    .. opt:: Firstorder Solver @ltac_expr
 
-      The default tactic used by :tacn:`firstorder` when no rule applies in
-      :g:`auto with core`.  It can be set locally or globally using this :term:`option`.
+      The default tactic used by :tacn:`firstorder` when no rule
+      applies in :g:`auto with core`. This command supports the same
```
Comment (jfehrle) [inline doc/sphinx/proofs/automatic-tactics/logic.rst:128]:
```suggestion
      in :g:`auto with core` applies. This command supports the same
```

**Review by jfehrle — COMMENTED** [review]
wording suggestions

File: doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst (hunk 5 above)
Comment (SkySkimmer) [reply doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst:3]: It's impossible to know without looking at the plugin's source.

File: doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst (hunk 6 above)
Comment (SkySkimmer) [reply doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst:6]: In 
~~~coq
Module M.
Module N.
End N.
End M.
~~~
M is a surrounding module for N

Comment (SkySkimmer) [reply doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst:6]: could also be called parent module

File: doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst (hunk 5 above)
Comment (jfehrle) [reply doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst:3]: Then how about referring to `@ltac_expr` if that's what matters?

Comment (SkySkimmer) [reply doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst:3]: It is not a grammar thing

File: doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst (hunk 6 above)
Comment (jfehrle) [reply doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst:6]: How about "Settings using `global` or without a locality attribute are currently applied when any module containing it (recursively) is imported."?

File: doc/sphinx/addendum/program.rst (hunk 7 above)
Comment (SkySkimmer) [reply doc/sphinx/addendum/program.rst:283]: could be either without loss of generality

File: doc/sphinx/proofs/automatic-tactics/logic.rst (hunk 9 above)
Comment (SkySkimmer) [reply doc/sphinx/proofs/automatic-tactics/logic.rst:128]: unrelated to this pr

File: doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst (hunk 5 above)
Comment (jfehrle) [reply doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst:3]: Agreed, not a grammar thing.  IIUC "tactic options" would be clearer as "options whose value is a tactic".  That's what I was trying to get at with `@ltac_expr`.

Comment (SkySkimmer) [reply doc/changelog/07-vernac-commands-and-options/15274-tacopt.rst:3]: It's really something each plugin using this system needs to say in their own doc, see paramcoq overlay

**Comment by SkySkimmer** [comment]
Assignee?

File: doc/sphinx/addendum/program.rst [hunk 10]
```diff
@@ -273,10 +273,18 @@ optional tactic is replaced by the default one if not specified.
...
-   otherwise.
+   otherwise. :attr:`export` and :attr:`global` may be used together.
+
+   When :attr:`global` is used without :attr:`export` and when no
+   explicit locality is used outside sections, the meaning is
+   different from the usual meaning of :attr:`global`: the command's
+   effect persists after the current module is closed (as with the
+   usual :attr:`global`), but it is also reapplied when the module or
+   any of its parents is imported. This will change in a future
+   version.
```
Comment (Alizter) [inline doc/sphinx/addendum/program.rst:287]: And when global is used with export?

Comment (Alizter) [reply doc/sphinx/addendum/program.rst:287]: I'm a little confused that "global + export" here means "usual global", perhaps you can make that clearer in the doc.

Comment (SkySkimmer) [reply doc/sphinx/addendum/program.rst:287]: global+export = normal global + normal export
global without export = default = normal global + export in all surrounding modules

**Comment by Alizter** [comment]
@coqbot merge now

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
@coqbot merge now
